File: deliverables/instagram-serums/remove_bg_vision.m
#import <Foundation/Foundation.h>
#import <Vision/Vision.h>
#import <CoreImage/CoreImage.h>

int main(int argc, const char *argv[]) {
  @autoreleasepool {
    if (argc != 3) return 1;
    NSURL *inputURL = [NSURL fileURLWithPath:[NSString stringWithUTF8String:argv[1]]];
    NSURL *outputURL = [NSURL fileURLWithPath:[NSString stringWithUTF8String:argv[2]]];
    CIImage *input = [CIImage imageWithContentsOfURL:inputURL];
    if (!input) return 2;
    VNGenerateForegroundInstanceMaskRequest *request = [VNGenerateForegroundInstanceMaskRequest new];
    VNImageRequestHandler *handler = [[VNImageRequestHandler alloc] initWithCIImage:input options:@{}];
    NSError *error = nil;
    if (![handler performRequests:@[request] error:&error]) return 3;
    VNInstanceMaskObservation *observation = request.results.firstObject;
    if (!observation) return 4;
    CVPixelBufferRef buffer = [observation generateScaledMaskForImageForInstances:observation.allInstances
                                                               fromRequestHandler:handler
                                                                            error:&error];
    if (!buffer) return 5;
    CIImage *mask = [CIImage imageWithCVPixelBuffer:buffer];
    CIImage *clear = [[CIImage imageWithColor:[CIColor colorWithRed:0 green:0 blue:0 alpha:0]] imageByCroppingToRect:input.extent];
    CIFilter *blend = [CIFilter filterWithName:@"CIBlendWithMask"];
    [blend setValue:input forKey:kCIInputImageKey];
    [blend setValue:clear forKey:kCIInputBackgroundImageKey];
    [blend setValue:mask forKey:kCIInputMaskImageKey];
    CIImage *result = blend.outputImage;
    CIContext *context = [CIContext contextWithOptions:nil];
    CGColorSpaceRef cs = CGColorSpaceCreateWithName(kCGColorSpaceSRGB);
    BOOL ok = [context writePNGRepresentationOfImage:result toURL:outputURL format:kCIFormatRGBA8 colorSpace:cs options:@{} error:&error];
    CGColorSpaceRelease(cs);
    CFRelease(buffer);
    return ok ? 0 : 6;
  }
}

File: deliverables/instagram-serums/remove_bg_vision.swift
import Foundation
import Vision
import CoreImage

guard CommandLine.arguments.count == 3 else { exit(1) }
let inputURL = URL(fileURLWithPath: CommandLine.arguments[1])
let outputURL = URL(fileURLWithPath: CommandLine.arguments[2])
guard let input = CIImage(contentsOf: inputURL) else { exit(2) }

let request = VNGenerateForegroundInstanceMaskRequest()
let handler = VNImageRequestHandler(ciImage: input)
try handler.perform([request])
guard let observation = request.results?.first else { exit(3) }
let maskBuffer = try observation.generateScaledMaskForImage(
  forInstances: observation.allInstances,
  from: handler
)
let mask = CIImage(cvPixelBuffer: maskBuffer)
let clear = CIImage(color: .clear).cropped(to: input.extent)
let result = input.applyingFilter("CIBlendWithMask", parameters: [
  kCIInputBackgroundImageKey: clear,
  kCIInputMaskImageKey: mask
])
let context = CIContext(options: [.useSoftwareRenderer: false])
let colorSpace = CGColorSpace(name: CGColorSpace.sRGB)!
try context.writePNGRepresentation(of: result, to: outputURL, format: .RGBA8, colorSpace: colorSpace)
